core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactory.js

```diff
@@ -0,0 +1,255 @@
+// Copyright 2018 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Factory for creating instances of NumberWithUnits and Units
+ * domain objects.
+ */
+
+oppia.constant('NUMBER_WITH_UNITS_PARSING_ERRORS', {
+  INVALID_VALUE:
+    'Please ensure that value is either a fraction or a number',
+  INVALID_CURRENCY:
+    'Please enter a valid currency (e.g., $ 5)',
+  INVALID_UNIT_CHARS:
+    'Please ensure that unit only contains numbers, alphabets, (, ), *, ^, /, -'
+});
+
+oppia.factory('NumberWithUnitsObjectFactory', [
+  'UnitsObjectFactory', 'FractionObjectFactory',
+  'NUMBER_WITH_UNITS_PARSING_ERRORS', function(
+      UnitsObjectFactory, FractionObjectFactory,
+      NUMBER_WITH_UNITS_PARSING_ERRORS) {
+    var NumberWithUnits = function(type, real, fractionObj, unitsObj) {
+      this.type = type;
+      this.real = real;
+      this.fraction = fractionObj;
+      this.units = unitsObj;
+    };
+
+    NumberWithUnits.prototype.toString = function() {
+      var numberWithUnitsString = '';
+      if (this.units.toString().includes('$')) {
+        this.units.units = this.units.toString().replace('$', '');
+        numberWithUnitsString += '$' + ' ';
+      }
+      if (this.units.toString().includes('Rs')) {
+        this.units.units = this.units.toString().replace('Rs', '');
+        numberWithUnitsString += 'Rs' + ' ';
+      }
+
+      if (this.type === 'real') {
+        numberWithUnitsString += this.real + ' ';
+      } else {
+        numberWithUnitsString += this.fraction.toString() + ' ';
+      }
+      numberWithUnitsString += this.units.toString();
+      numberWithUnitsString = numberWithUnitsString.trim();
+
+      return numberWithUnitsString;
+    };
+
+    NumberWithUnits.fromRawInputString = function(rawInput) {
+      rawInput = rawInput.trim();
+      var type = '';
+      var real = '';
+      var fractionObj = '';
+      var units = '';
+      var value = 0;
+      var unitObj = '';
+
+      // Start with digit when there is no currency unit.
+      if (rawInput.match(/^\d/)) {
+        var ind = rawInput.indexOf(rawInput.match(/[a-z(]/i));
+        if (ind === -1) {
+          // There is value with no units.
+          value = rawInput;
+          units = '';
+        } else {
+          value = rawInput.substr(0, ind).trim();
+          units = rawInput.substr(ind).trim();
+        }
+      } else {
+        var ind = rawInput.indexOf(' ');
```

Comment:
Use a simple regex here to match `/(\$\s|Rs\s|₹\s)/` and find index like that.

Reply:
I'm doing validation on currencies also. In this case, I need to change the whole lot of code. I think what I implemented is also correct as far as I'm validating currency units.

Reply:
Done. Now, currency units are validated explicitly. Added the test case too.